crates/ruma-events/src/room/member.rs

```diff
@@ -120,6 +122,89 @@ impl RoomMemberEventContent {
     }
 }
 
+impl RedactContent for RoomMemberEventContent {
+    type Redacted = RedactedRoomMemberEventContent;
+
+    fn redact(self, _version: &RoomVersionId) -> RedactedRoomMemberEventContent {
+        RedactedRoomMemberEventContent {
+            membership: self.membership,
+            #[cfg(feature = "unstable-pre-spec")]
+            join_authorised_via_users_server: match _version {
+                #[cfg(feature = "unstable-pre-spec")]
+                RoomVersionId::V9 => self.join_authorised_via_users_server,
+                _ => None,
+            },
+        }
+    }
+}
+
+/// A member event that has been redacted.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct RedactedRoomMemberEventContent {
+    /// The membership state of this user.
+    pub membership: MembershipState,
+
+    /// An arbitrary user who has the power to issue invites.
+    ///
+    /// This is redacted in room versions 8 and below. It is used for validating
+    /// joins when the join rule is restricted.
+    #[cfg(feature = "unstable-pre-spec")]
+    pub join_authorised_via_users_server: Option<Box<UserId>>,
+}
+
+impl RedactedRoomMemberEventContent {
+    /// Create a `RedactedRoomMemberEventContent` with the given membership.
+    ///
+    /// This is only valid for room version 8 and below.
+    pub fn new_v1(membership: MembershipState) -> Self {
+        Self {
+            membership,
+            #[cfg(feature = "unstable-pre-spec")]
+            join_authorised_via_users_server: None,
+        }
+    }
+
+    /// Create a `RedactedRoomMemberEventContent` with the given membership and auth_user.
+    ///
+    /// This is only valid for room version 9 and above.
+    #[cfg(feature = "unstable-pre-spec")]
+    pub fn new_v9(membership: MembershipState, auth_user: Box<UserId>) -> Self {
```

Comment:
Does this constructor really make sense? Not every redacted room member event in a v9+ room has an auth_user, right? I think we should just have a single constructor `new`, and let people assign `join_authorised_via_users_server` separately.